src/smc-webapp/assistant/header.cjsx

```diff
@@ -0,0 +1,164 @@
+    display        : 'flex'
+    flexWrap       : 'nowrap'
+    justifyContent : 'space-between'
+
+inner_style = immutable.Map
+    marginLeft  : '10px'
+    marginRight : '10px'
+
+title_style  = inner_style.merge({flex: '0 0 auto'}).toJS()
+nav_style    = inner_style.merge({flex: '1 0 auto'}).toJS()
+search_style = inner_style.merge({flex: '0 1 auto', marginRight: '50px'}).toJS()
+
+exports.ExamplesHeader = rclass
+    displayName : 'ExamplesHeader'
+
+    propTypes:
+        actions      : rtypes.object
+        nav_entries  : rtypes.arrayOf(rtypes.string)
+        search_str   : rtypes.string
+        lang         : rtypes.string
+        lang_select  : rtypes.bool
+        unknown_lang : rtypes.bool
+
+    getDefaultProps: ->
+        search_str : ''
+
+    shouldComponentUpdate: (props, state) ->
+        ret = misc.is_different(@props, props, ['lang', 'search_str', 'search_sel', 'unknown_lang'])
+        ret or= misc.is_different_array(props.nav_entries, @props.nav_entries)
+        return ret
+
+    langSelect: (key) ->
+        @props.actions.select_lang(key)
+
+    search: (evt) ->
+        evt.preventDefault()
+        evt.stopPropagation()
+        @props.actions.search(evt.target.value)
+
+    handle_search_keyup: (evt) ->
+        if not @props.search_str?.length
+            return true
+        switch evt.keyCode
+            when 27 # ESC
+                if @props.search_str?.length > 0
+                    @props.actions.search('')
+                else
+                    return true
+            when 38
+                @props.actions.search_cursor(-1)
+            when 40
+                @props.actions.search_cursor(+1)
+            else
+                # let them propagate up to the dialog's key handler
+                return true
+        evt.preventDefault() # which
+        evt.stopPropagation() # does
+        evt.nativeEvent.stopImmediatePropagation() # what ?!
+        return false
+
+    search_clear: ->
+        @props.actions.search('')
+
+    # a horizontal list of clickable language categories
+    render_nav: ->
+        entries = @props.nav_entries ? []
+        {lang2name} = require('./main')
+
+        <Nav
+            bsStyle   = {'pills'}
+            style     = {marginLeft:'10px'}
+            activeKey = {@props.lang}
+            ref       = {'lang'}
+            onSelect  = {@langSelect}
+        >
+            {
+                entries.map (key, idx) ->
+                    name = lang2name(key)
+                    <NavItem
+                        key      = {idx}
+                        eventKey = {key}
+                        title    = {name}
+                    >
+                        {name}
+                    </NavItem>
+            }
+        </Nav>
+
+    render_search: ->
+        bsStyle = if @props.search_str?.length > 0 then 'warning' else 'default'
+        <FormGroup>
+            <InputGroup className = {'webapp-examples-search'}>
+                <FormControl
```

Comment:
Please use `SearchInput` from r_misc. It's semantically very nice. I don't see anything here at first glance that would warrant rolling your own.